rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -32,491 +33,347 @@
-    return false;
-  }
+      const liveliness::TopicInfo topic_info = entity.topic_info().value();
+      GraphNode::TopicMap & topic_map = entity.type() ==
+        EntityType::Publisher ? graph_node.pubs_ : graph_node.subs_;
+      const std::string entity_desc = entity.type() ==
+        EntityType::Publisher ? "publisher" : "subscription";
+      const std::size_t pub_count = entity.type() == EntityType::Publisher ? 1 : 0;
+      const std::size_t sub_count = !pub_count;
+      TopicDataPtr graph_topic_data = std::make_shared<TopicData>(
+        topic_info,
+        TopicStats{pub_count, sub_count});
+
+      GraphNode::TopicDataMap topic_data_map = {
+        {graph_topic_data->info_.type_, graph_topic_data}};
+      std::pair<GraphNode::TopicMap::iterator, bool> insertion =
+        topic_map.insert(std::make_pair(topic_info.name_, topic_data_map));
+      if (!insertion.second) {
+        // A topic with the same name already exists in the node so we append the type.
+        std::pair<GraphNode::TopicDataMap::iterator, bool> type_insertion =
+          insertion.first->second.insert(
+          std::make_pair(
+            graph_topic_data->info_.type_,
+            graph_topic_data));
+        if (!type_insertion.second) {
+          // We have another instance of a pub/sub over the same topic and type so we increment
+          // the counters.
+          TopicDataPtr & existing_graph_topic = type_insertion.first->second;
+          existing_graph_topic->stats_.pub_count_ += pub_count;
+          existing_graph_topic->stats_.sub_count_ += sub_count;
+        }
+      }
 
-  // TODO(Yadunund): z_keyexpr_new creates a copy so find a way to avoid it.
-  z_owned_keyexpr_t keyexpr = z_keyexpr_new(token.c_str());
-  auto drop_keyexpr = rcpputils::make_scope_exit(
-    [&keyexpr]() {
-      z_drop(z_move(keyexpr));
-    });
-  if (!z_keyexpr_check(&keyexpr)) {
-    RCUTILS_SET_ERROR_MSG("invalid keyexpression generation for liveliness publication.");
-    return false;
-  }
-  RCUTILS_LOG_WARN_NAMED("rmw_zenoh_cpp", "Sending PUT on %s", token.c_str());
-  z_put_options_t options = z_put_options_default();
-  options.encoding = z_encoding(Z_ENCODING_PREFIX_EMPTY, NULL);
-  if (z_put(z_loan(*session), z_keyexpr(token.c_str()), nullptr, 0, &options) < 0) {
-    RCUTILS_SET_ERROR_MSG("unable to publish liveliness for node creation");
-    return false;
-  }
+      // Bookkeeping: Update graph_topics_ which keeps track of topics across all nodes in the graph
+      GraphNode::TopicMap::iterator cache_topic_it = graph_topics_.find(topic_info.name_);
+      if (cache_topic_it == graph_topics_.end()) {
+        // First time this topic name is added to the graph.
+        auto topic_data_ptr = std::make_shared<TopicData>(
+          topic_info,
+          TopicStats{pub_count, sub_count}
+        );
+        graph_topics_[topic_info.name_] = GraphNode::TopicDataMap{
+          {topic_info.type_, topic_data_ptr}
+        };
```

Comment:
Isn't it the same as `topic_data_map`?

```suggestion
        graph_topics_[topic_info.name_] = topic_data_map;
```

Reply:
Done https://github.com/ros2/rmw_zenoh/pull/70/commits/357f6588fab1f2f9192e7be418d9f1c3b4956da5